### PR TITLE
Defines an interface for the host, to allow the node API to be uncoupled

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

The node's RPC API should be in its own package, but this was impossible due to a circular dependency caused by the fact that there was no interface for the host class.

### What changes were made as part of this PR:

Refactoring.

- Introduces a `Host` interface
- Pulls subpackages in the `host` package into a more sensible structure

### CODE REVIEW

File: go/host/interfaces.go [hunk 1]
```diff
@@ -4,17 +4,42 @@ import (
...
 	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/config"
+	"github.com/obscuronet/go-obscuro/go/ethadapter"
+	"github.com/obscuronet/go-obscuro/go/host/db"
 )
 
-// P2PCallback -the glue between the P2p layer and the node. Notifies the node when rollups and transactions are received from peers
-type P2PCallback interface {
+// Host is the half of the Obscuro node that lives outside the enclave.
+type Host interface {
```
Comment (tudor-malene) [inline go/host/interfaces.go:13]: chatted on discord. It woudl be good to break this up in 2 interfaces: "Host" and "MockHost"

**Review by tudor-malene — CHANGES_REQUESTED** [review]
looks good in principle.
A bit more polish needed .

File: go/host/node/node.go [hunk 2]
```diff
@@ -34,18 +39,17 @@ const (
 	apiNamespaceNetwork  = "net"
 )
 
-// Node this will become the Obscuro "Node" type
+// Node is an implementation of hostinterfaces.Host.
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/node/node.go:42]: Fixed.

File: go/host/node/node.go [hunk 3]
```diff
@@ -57,8 +61,7 @@ type Node struct {
 	rollupsP2PCh chan common.EncodedRollup  // The channel that new rollups from peers are sent to
 	txP2PCh      chan common.EncryptedTx    // The channel that new transactions from peers are sent to
 
-	nodeDB       *DB    // Stores the node's publicly-available data
-	readyForWork *int32 // Whether the node has bootstrapped the existing blocks and has the enclave secret
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/node/node.go:61]: Apparently not!

File: go/host/interfaces.go [hunk 4]
```diff
@@ -4,17 +4,42 @@ import (
...
+	// ReceiveTx processes a transaction received from a peer host.
 	ReceiveTx(tx common.EncryptedTx)
+	// Stop gracefully stops the host execution.
+	Stop()
+
+	/// The following methods are only used for integration testing.
+
+	P2P() P2P
+	// MockedNewHead receives the notification of new blocks.
+	MockedNewHead(b common.EncodedBlock, p common.EncodedBlock)
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/interfaces.go:36]: Added.

File: go/host/interfaces.go (hunk 1 above)
Comment (joeldudley) [reply go/host/interfaces.go:13]: Done.

File: go/host/rpc/clientapi/client_api_eth.go [hunk 5]
```diff
@@ -1,11 +1,13 @@
-package host
+package clientapi
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/host/rpc/clientapi/client_api_eth.go:1]: I'll update the main README. Which other README are you referring to?

[the remark this answers is not in the document]
Comment (tudor-malene) [reply go/host/rpc/clientapi/client_api_eth.go:1]: ah, I thought there is one in the rpc folder. It was actually in the parent folder. 
Ignore.

**Review by tudor-malene — APPROVED** [review]
lgtm

**Pull request merged**